File: project/two_pianos/lib/two_pianos_web/channels/user_channel.ex
defmodule TwoPianosWeb.UserChannel do

  use TwoPianosWeb, :channel

  # User joins this channel to be accessbile by an external broadcast to "user:<user_id>"

  def join("user:" <> user_id, _, socket) do
    {:ok, socket}
  end

  def broadcast_match(user_id, room_id) do
    user_id_str = user_id |> Atom.to_string()
    TwoPianosWeb.Endpoint.broadcast("user:" <> user_id_str, "matched_user", %{room_id: room_id})
    :ok
  end

end

File: project/two_pianos/assets/js/user_channel.js
export class UserChannel{

    static setup_and_join(socket, user_id){

        let user_channel = socket.channel(`user:${user_id}`)

        // when another user connects to you after you hit "join stranger" button
        user_channel.on("matched_user", (msg) => {

            console.log(`Matched! ${msg.room_id}`)

            localStorage.setItem("room_id", msg.room_id)

            setTimeout(function(){
                window.location.replace("http://localhost:4000/room")
            }, 2000)
            
        })

        user_channel.join()

        return user_channel

    }

}